### ISSUE TITLE
Access Custom Metadata via API

### ISSUE THREAD

**Issue by DjTuxeedo** [lachlan-00/python3-ampache#4]
Hey, 
I was wondering: 
Is it possible to access custom Metadata of a song via python-ampache?
I enabled bpm as a custom Metadata field. It works in the gui-interface and bpm values are displayed in the song details card. So the values must be stored somewhere in the ampache database.
Do you think it might be possible to add these custom metadata fields in the output of:

playlist_songs = ampache.playlist_songs(ampache_url, ampache_api, single_playlist)

? That would be awesome :-)

**Comment by lachlan-00** [comment]
do you mean extra data for a song item like this one?

I can add the metadata if it exists, i've never used it myself but can you put some screens of the extra stuff?
```
<song id="164215">
	<title><![CDATA[!]]></title>
	<name><![CDATA[!]]></name>
	<artist id="8526"><![CDATA[Hellyeah]]></artist>
	<album id="56038"><![CDATA[Unden!able]]></album>
	<tag id="20" count="1" ><![CDATA[Metal]]></tag>
	<tag id="37" count="1" ><![CDATA[Hard Rock]]></tag>
	<filename><![CDATA[/mnt/music/Hellyeah/(2016) Unden!able [ESM 171-2]/101 - !.mp3]]></filename>
	<track>1</track>
	<time>79</time>
	<year>2016</year>
	<bitrate>320000</bitrate>
	<rate>44100</rate>
	<mode>cbr</mode>
	<mime>audio/mpeg</mime>
	<url><![CDATA[https://music.com.au/play/index.php?ssid=eeb9f1b6056246a7d563f479f518bb34&type=song&oid=164215&uid=2&player=api&name=Hellyeah%20-%20-.mp3]]></url>
	<size>3184944</size>
	<mbid>42bb2b7a-bf67-4cc4-8610-25399d56474b</mbid>
	<album_mbid></album_mbid>
	<artist_mbid></artist_mbid>
	<albumartist_mbid></albumartist_mbid>
	<art><![CDATA[https://music.com.au/image.php?object_id=56038&object_type=album&auth=eeb9f1b6056246a7d563f479f518bb34&name=art.jpg]]></art>
	<flag>0</flag>
	<preciserating>0</preciserating>
	<rating>0</rating>
	<averagerating>0</averagerating>
	<composer><![CDATA[]]></composer>
	<channels></channels>
	<comment><![CDATA[]]></comment>
	<publisher><![CDATA[Eleven Seven Music]]></publisher>
	<language></language>
	<replaygain_album_gain>0.000000</replaygain_album_gain>
	<replaygain_album_peak>0.000000</replaygain_album_peak>
	<replaygain_track_gain>-6.920000</replaygain_track_gain>
	<replaygain_track_peak>1.000000</replaygain_track_peak>
	<genre><![CDATA[Metal]]></genre>
	<genre><![CDATA[Hard Rock]]></genre>
</song>
```

**Comment by Rhythm-of-Time** [comment]
In the current UI, I can't see this feature working. Under System, I set the Custom Metadata field to "isrc, gema". By entering these strings I expected some new text fields to appear in a dialog.

They neither appear in the song nor album edit dialog. Where could I see them, apart from the configuration entry in the database? Or rather, is this exclusively for tracks that already _contain_ this value in the ID3 tags?

**Comment by lachlan-00** [comment]
I forget how to set it up. but they need to be in the disable list as well to be "enabled" if i remember correctly

**Comment by lachlan-00** [comment]
there's the config option as well. I think this needs a writeup in the wiki as i really forget this completely.

```
; Enable importing custom metadata from files.
; This will need a bit of time during the import. So you may want to disable this
; if you have troubles with huge databases.
; DEFAULT: false
;enable_custom_metadata = "false"
```

**Comment by DjTuxeedo** [comment]
Yes, the Metadata I'm most interested in as a DJ is the "BPM" - value. Ampache is reading the value just fine, as I can see from the song - Details dialogue. Accessing it from pythonAmpache would be awesome and Displayling it in the Library/Playlist view and maybe even sorting by that value would put Ampache miles ahead in my opinion :-)

**Comment by lachlan-00** [comment]
That shouldn't be hard. The data is all there in the db and it's probably already possible with advanced_search.

I also think it would make a good playlist_generate option

**Comment by DjTuxeedo** [comment]
I like the way you're thinking :-)
And I wish there was some way I could help you with any of this.

**Comment by DjTuxeedo** [comment]
ups

**Comment by lachlan-00** [comment]
That's what we're here for.
Happy to work on this for the next API release. It will also help us document the custom metadata

**Comment by DjTuxeedo** [comment]
by the way, is there newsletter I could subscribe to? So I wont miss the next release?

**Comment by lachlan-00** [comment]
@Rhythm-of-Time enabling the custom_meta_data option looks to be all that's needed. it adds a more data field in the edit window
![image](https://user-images.githubusercontent.com/1305249/73051037-7461b680-3ecd-11ea-817f-0d90d8614d98.png)

which lists all my extra tags.
![image](https://user-images.githubusercontent.com/1305249/73051087-99562980-3ecd-11ea-8ceb-9600b049b6d6.png)

I don't really know why this isn't the default but it also means don't need to check much more here.

@DjTuxeedo from what i can see all you need to do is advanced_search. I'll work it out in an example but advanced search is actually just a copy of the search ui. (below screen) This should be easy enough to document and then we will be able to simplify it further from there.
![image](https://user-images.githubusercontent.com/1305249/73051267-17b2cb80-3ece-11ea-8ae7-d6823281a7dc.png)

**Comment by lachlan-00** [comment]
> by the way, is there newsletter I could subscribe to? So I wont miss the next release?

Ampache now has, twitter, mastodon, telegram, reddit and google groups to choose from. we'll announce things from probably all of these now that it's not just me

**Comment by Rhythm-of-Time** [comment]
enable_custom_metadata was set in config, but clicking on More Metadata shows only the first two items (Encoding and Audio Codec) from the disable list, but none of my Define Field List. How does disabling items work? In the disable list I see only 8 tags, not all of the ones shown above in your example. Am I supposed to mark the desired fields and then click "Update Preferences"? Nothing happens if I do that... And how do I reset this list to activate all fields?

Did I mention that the files are all WAV, which do not contain any tags?  Mp3 with a complete set of ID3 tags yield a adifferent result. It looks like properties that are _not_ _present_ in the file _cannot be set_ from the edit dialog, because they don't appear in the dialog. How to attach this information to WAV files, then? I don't like BWF, because itthey tend to be problematic. So I'd like to put this information only into the database (and not the file itself) - ideally by reading in an XLS / CSV document containing the metadata of all tracks.

It looks like using other formats than mp3 really requires some work. Well, it should be possible by using the API... isn't it?

**Comment by lachlan-00** [comment]
Ive been through the code and it's implementation is confising.
It looks like the metadata feature was written by someone in the project I've not heard of before. It also seems to only be for songs but we should allow the API to apply to all songs in an album and artist

To get all the extra tag data I had to verify my files. I'm doing it for my whole catalogue now.

But I think we should be allowing.
* Definition of extra fields (UI+API)
* Set fields for songs based on search/groups (API)
* These extra fields to always be available in the edit window.
* A default list in the database so we can at least define a list for the api

**Comment by Rhythm-of-Time** [comment]
How about moving / copying this thread over to Ampache? It's a bit confusing to have this under python.

**Comment by lachlan-00** [comment]
I'll look at moving the repo over to the org, i can't transfer issues from my personal account to the org which is annoying.
